File: packages/sdk-node/src/approvals.ts
/**
 * RuneSignal Node SDK — Approvals (HITL Gateway) Resource
 * POST /api/v1/approval-requests
 */

import { BaseClient } from './client';
import { RuneSignalError } from './types';

export type BlastRadiusLevel = 'low' | 'medium' | 'high' | 'critical';
export type SlaAutoAction = 'approve' | 'reject' | 'escalate';
export type ApprovalStatus = 'pending' | 'approved' | 'rejected' | 'expired';

export interface BlastRadius {
  level?: BlastRadiusLevel;
  reversible: boolean;
  affectsExternalSystem?: boolean;
  dataClassification?: 'public' | 'internal' | 'confidential' | 'restricted';
  affectedRecordCount?: number;
  estimatedImpactUsd?: number;
}

export interface SubmitApprovalRequest {
  agentId: string;
  actionType: string;
  actionSummary: string;
  blastRadius: BlastRadius;
  payload?: Record<string, unknown>;
  slaHours?: number;
  slaAutoAction?: SlaAutoAction;
  idempotencyKey?: string;
  metadata?: Record<string, unknown>;
}

export interface ApprovalTicket {
  id: string;
  agentId: string;
  actionType: string;
  actionSummary: string;
  status: ApprovalStatus;
  slaDeadline?: string;
  slaAutoAction?: SlaAutoAction;
  reviewUrl?: string;
  resolvedBy?: string;
  resolvedAt?: string;
  resolutionReason?: string;
  idempotencyKey?: string;
  createdAt: string;
}

export interface PollOptions {
  intervalMs?: number;
  timeoutMs?: number;
}

export class ApprovalsResource {
  constructor(private readonly client: BaseClient) {}

  /**
   * Submit a HITL approval request.
   * Pass an idempotencyKey to make it safe to retry without creating duplicates.
   */
  async submit(request: SubmitApprovalRequest): Promise<ApprovalTicket> {
    const headers: Record<string, string> = {};
    if (request.idempotencyKey) {
      headers['Idempotency-Key'] = request.idempotencyKey;
    }

    const raw: any = await (this.client as any).request('POST', '/api/v1/approval-requests', {
      body: {
        agent_id: request.agentId,
        action_type: request.actionType,
        action_summary: request.actionSummary,
        blast_radius: {
          level: request.blastRadius.level,
          reversible: request.blastRadius.reversible,
          affects_external_system: request.blastRadius.affectsExternalSystem,
          data_classification: request.blastRadius.dataClassification,
          affected_record_count: request.blastRadius.affectedRecordCount,
          estimated_impact_usd: request.blastRadius.estimatedImpactUsd,
        },
        payload: request.payload,
        sla_hours: request.slaHours,
        sla_auto_action: request.slaAutoAction,
        idempotency_key: request.idempotencyKey,
        metadata: request.metadata,
      },
    });

    return mapTicket(raw);
  }

  /**
   * Poll for a decision on a pending approval ticket.
   * Resolves when status changes from 'pending' or timeout is reached.
   */
  async poll(ticketId: string, options: PollOptions = {}): Promise<ApprovalTicket> {
    const { intervalMs = 5_000, timeoutMs = 300_000 } = options;
    const deadline = Date.now() + timeoutMs;

    while (Date.now() < deadline) {
      const ticket = await this.get(ticketId);
      if (ticket.status !== 'pending') return ticket;
      await sleep(intervalMs);
    }

    throw new RuneSignalError(
      `Approval poll timed out after ${timeoutMs}ms for ticket ${ticketId}`,
      408,
      'APPROVAL_POLL_TIMEOUT'
    );
  }

  /**
   * Submit and wait for a decision in one call.
   * Polls automatically at 5s intervals.
   */
  async requestApproval(
    request: SubmitApprovalRequest,
    pollOptions: PollOptions = {}
  ): Promise<ApprovalTicket> {
    const ticket = await this.submit(request);
    return this.poll(ticket.id, pollOptions);
  }

  /** Get a ticket by ID. */
  async get(ticketId: string): Promise<ApprovalTicket> {
    const raw: any = await (this.client as any).request('GET', `/api/v1/approval-requests/${ticketId}`);
    return mapTicket(raw);
  }

  /**
   * Manually resolve a ticket (approve or reject).
   * Used by review apps and dashboard integrations.
   */
  async resolve(
    ticketId: string,
    decision: { action: 'approve' | 'reject'; reviewerId: string; reason?: string }
  ): Promise<ApprovalTicket> {
    const raw: any = await (this.client as any).request(
      'POST',
      `/api/v1/approval-requests/${ticketId}/resolve`,
      {
        body: {
          action: decision.action,
          reviewer_id: decision.reviewerId,
          reason: decision.reason,
        },
      }
    );
    return mapTicket(raw);
  }

  /** List recent approval tickets. */
  async list(options: { status?: ApprovalStatus; limit?: number; agentId?: string } = {}): Promise<ApprovalTicket[]> {
    const query: Record<string, string> = {};
    if (options.status)  query.status   = options.status;
    if (options.limit)   query.limit    = String(options.limit);
    if (options.agentId) query.agent_id = options.agentId;

    const raw: any[] = await (this.client as any).request('GET', '/api/v1/approval-requests', { query });
    return raw.map(mapTicket);
  }
}

function mapTicket(raw: any): ApprovalTicket {
  return {
    id:                raw.id,
    agentId:           raw.agent_id,
    actionType:        raw.action_type,
    actionSummary:     raw.action_summary,
    status:            raw.status,
    slaDeadline:       raw.sla_deadline,
    slaAutoAction:     raw.sla_auto_action,
    reviewUrl:         raw.review_url,
    resolvedBy:        raw.resolved_by,
    resolvedAt:        raw.resolved_at,
    resolutionReason:  raw.resolution_reason,
    idempotencyKey:    raw.idempotency_key,
    createdAt:         raw.created_at,
  };
}

function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

File: packages/sdk-python/runesignal/approvals.py
"""RuneSignal Python SDK — Approvals (HITL Gateway)"""
from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import Any, Dict, List, Literal, Optional

from .http import BaseHTTPClient
from .types import RuneSignalError

ApprovalStatus = Literal["pending", "approved", "rejected", "expired"]
BlastRadiusLevel = Literal["low", "medium", "high", "critical"]
SlaAutoAction = Literal["approve", "reject", "escalate"]


@dataclass
class ApprovalTicket:
    id: str
    agent_id: str
    action_type: str
    action_summary: str
    status: str
    sla_deadline: Optional[str] = None
    sla_auto_action: Optional[str] = None
    review_url: Optional[str] = None
    resolved_by: Optional[str] = None
    resolved_at: Optional[str] = None
    resolution_reason: Optional[str] = None
    idempotency_key: Optional[str] = None
    created_at: str = ""


@dataclass
class SubmitApprovalRequest:
    agent_id: str
    action_type: str
    action_summary: str
    blast_radius: Dict[str, Any]
    payload: Optional[Dict[str, Any]] = None
    sla_hours: Optional[int] = None
    sla_auto_action: Optional[str] = None
    idempotency_key: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None


def _map_ticket(raw: Dict[str, Any]) -> ApprovalTicket:
    return ApprovalTicket(
        id=raw["id"],
        agent_id=raw.get("agent_id", ""),
        action_type=raw.get("action_type", ""),
        action_summary=raw.get("action_summary", ""),
        status=raw.get("status", "pending"),
        sla_deadline=raw.get("sla_deadline"),
        sla_auto_action=raw.get("sla_auto_action"),
        review_url=raw.get("review_url"),
        resolved_by=raw.get("resolved_by"),
        resolved_at=raw.get("resolved_at"),
        resolution_reason=raw.get("resolution_reason"),
        idempotency_key=raw.get("idempotency_key"),
        created_at=raw.get("created_at", ""),
    )


class ApprovalsResource:
    def __init__(self, http: BaseHTTPClient):
        self._http = http

    async def submit(self, request: SubmitApprovalRequest) -> ApprovalTicket:
        raw = await self._http.request(
            "POST",
            "/api/v1/approval-requests",
            body={
                "agent_id": request.agent_id,
                "action_type": request.action_type,
                "action_summary": request.action_summary,
                "blast_radius": request.blast_radius,
                "payload": request.payload,
                "sla_hours": request.sla_hours,
                "sla_auto_action": request.sla_auto_action,
                "idempotency_key": request.idempotency_key,
                "metadata": request.metadata,
            },
            idempotency_key=request.idempotency_key,
        )
        return _map_ticket(raw)

    async def poll(
        self,
        ticket_id: str,
        interval_seconds: float = 5.0,
        timeout_seconds: float = 300.0,
    ) -> ApprovalTicket:
        import time
        deadline = time.monotonic() + timeout_seconds
        while time.monotonic() < deadline:
            ticket = await self.get(ticket_id)
            if ticket.status != "pending":
                return ticket
            await asyncio.sleep(interval_seconds)
        raise RuneSignalError(
            f"Approval poll timed out after {timeout_seconds}s for ticket {ticket_id}",
            408,
            "APPROVAL_POLL_TIMEOUT",
        )

    async def request_approval(
        self,
        agent_id: str,
        action_type: str,
        action_summary: str,
        blast_radius: Dict[str, Any],
        payload: Optional[Dict[str, Any]] = None,
        sla_hours: Optional[int] = None,
        sla_auto_action: Optional[str] = None,
        idempotency_key: Optional[str] = None,
        poll_interval: float = 5.0,
        poll_timeout: float = 300.0,
    ) -> ApprovalTicket:
        ticket = await self.submit(
            SubmitApprovalRequest(
                agent_id=agent_id,
                action_type=action_type,
                action_summary=action_summary,
                blast_radius=blast_radius,
                payload=payload,
                sla_hours=sla_hours,
                sla_auto_action=sla_auto_action,
                idempotency_key=idempotency_key,
            )
        )
        return await self.poll(ticket.id, poll_interval, poll_timeout)

    async def get(self, ticket_id: str) -> ApprovalTicket:
        raw = await self._http.request("GET", f"/api/v1/approval-requests/{ticket_id}")
        return _map_ticket(raw)

    async def resolve(
        self,
        ticket_id: str,
        action: Literal["approve", "reject"],
        reviewer_id: str,
        reason: Optional[str] = None,
    ) -> ApprovalTicket:
        raw = await self._http.request(
            "POST",
            f"/api/v1/approval-requests/{ticket_id}/resolve",
            body={"action": action, "reviewer_id": reviewer_id, "reason": reason},
        )
        return _map_ticket(raw)

    async def list(
        self,
        status: Optional[str] = None,
        agent_id: Optional[str] = None,
        limit: Optional[int] = None,
    ) -> List[ApprovalTicket]:
        query: Dict[str, str] = {}
        if status:   query["status"]   = status
        if agent_id: query["agent_id"] = agent_id
        if limit:    query["limit"]    = str(limit)
        raw_list = await self._http.request("GET", "/api/v1/approval-requests", query=query)
        return [_map_ticket(r) for r in raw_list]
